### ISSUE TITLE
Is there a way to reduce WAL file size under pg_wal directory for clusters?

### ISSUE THREAD

**Issue by m0sh1x2** [zalando/postgres-operator#1743]
Hello,

I am using the base minimal cluster with one master and one worker node and using the v1.7.1 version of the operator with default settings.

Currently I am noticing that the pg_wal file directory grows quite a lot in some cases and eats up space really fast from several GB per day without cleanup.

Is there a way to reduce WAL files on the minimal cluster or this is an expected functionality.

As I understand backing up the cluster should automatically clear the WAL files but sometimes they stay for days and in other cases they fill up the storage for example for 1-2GB increase per day for a database with size ~400MB.

Please let me know if I might be missing something or the size of the WAL files is expected to grow in time.

Thanks

----------------
I have also answered the required questions for debugging:

Please, answer some short questions which should help us to understand your problem / question better?

- **Which image of the operator are you using?** e.g. registry.opensource.zalan.do/acid/postgres-operator:v1.7.1

- **Where do you run it - cloud or metal? Kubernetes or OpenShift?** [
minikube, kind and op-prem install with kubeadm

- **Are you running Postgres Operator in production?** 
- yes

- **Type of issue?** [Bug report, question, feature request, etc.]
question

Some general remarks when posting a bug report:
- Please, check the operator, pod (Patroni) and postgresql logs first. When copy-pasting many log lines please do it in a separate GitHub gist together with your Postgres CRD and configuration manifest.
- If you feel this issue might be more related to the [Spilo](https://github.com/zalando/spilo/issues) docker image or [Patroni](https://github.com/zalando/patroni/issues), consider opening issues in the respective repos.

**Comment by DaspawnW** [comment]
For us the same. We have an instance where the extracted backup was 2,3 MB but the 5 GB PVC was full.

**Comment by thangamani-arun** [comment]
We are also facing similar issues. When I dig into the pods, found that `/home/postgres/pgdata/pgroot/data/pg_wal` directory has lot for wal log files.

The wal log files were not cleaned, hence the logs file fills the pod disk space and postgresql will not respond anymore.
any way to cleanup the files as it is archiving over remote S3 ?

**Comment by nitindamle** [comment]
we are using kubernates cluster on top of private cloud,we are also facing similar issues. When I dig into the pods, found that /home/postgres/pgdata/pgroot/data/pg_wal directory has lot for wal log files with GB's and manytime pod gets crash due to no space left on disk.

The wal log files were not cleaned, hence the logs file fills the pod disk space and postgresql will not respond anymore.
any way to cleanup the files as it is archiving over remote S3 ?

**Comment by tydra-wang** [comment]
+1

**Comment by CyberDem0n** [comment]
There are more or less three reasons why pg_wal is growing:
1. checkpoints not happening (very unlikely)
2. unused replication slot
3. failing `archive_command`

You need to investigate, find the reason, and eliminate the problem.
The starting point would be Postgres logs located in `$PGDATA/../pg_log`, `SELECT * FROM pg_replication_slots`, and `ps auxwf` output.

**Comment by tydra-wang** [comment]
How can I figure out the max disk size pg_wal would cost?

**Comment by spreeker** [comment]
Hi for anyone landing here figureing out why their database cluster is running out of space.

executing `du -h -d 4` showed me that
the wal folder got realy large  ` ./pgdata/pgroot/data/pg_wal`. The reason for that is because replication nodes where not longer healthy and catching up. 

I solvend it by executing on a node:  `patronictl reinit`   and selecting the unhealty replicas.  You can see the status of replicas with `patronictl list`.  When the nodes where healty again my wal folder size when from 144gb to near zero on the database nodes.

**Comment by thangamani-arun** [comment]
@spreeker : I agree that procedure works. But since these container, unless you monitor the disk usage and cluster status you will not be able to take such actions. It may lead to data loss as well.

**Comment by spreeker** [comment]
Yep very inconvenient.

**Comment by jenishp619** [comment]
Hello,

I've been exploring PostgreSQL's pg_wal, archiving, and recovery mechanisms, and I'd like to share a use case scenario from my project.  I have logical replication set up between two VMs, each running PostgreSQL 16.

I've developed a custom bash script that uses pg_controldata to identify the oldest WAL file currently in use by PostgreSQL.  This file is considered the oldest that the system requires.  The script also checks if any replication slots require specific WAL files, as subscribers lagging behind might need these files to catch up.

The script retrieves a list of all WAL files required by existing replication slots.  It then applies the following logic:  If a file is required by a replication slot, it is not deleted. Otherwise, if the file is not used by a replication slot and is older than the oldest WAL file currently in use by the PostgreSQL system, it is deleted.  For context, archiving is enabled, and all WAL files from pg_wal/* are copied to an archive/ directory for recovery and replication purposes.

PostgreSQL 16 also offers the recovery_command = 'command' option. This allows PostgreSQL to retrieve WAL file information from the archive/ directory if a WAL file is deleted from pg_wal before being consumed by a subscriber.

Please feel free to contact me if you have any questions about this concept or if you'd like assistance with your specific use case. Happy coding!

**Comment by jenishp619** [comment]
> Hello,
> 
> I've been exploring PostgreSQL's pg_wal, archiving, and recovery mechanisms, and I'd like to share a use case scenario from my project. I have logical replication set up between two VMs, each running PostgreSQL 16.
> 
> I've developed a custom bash script that uses pg_controldata to identify the oldest WAL file currently in use by PostgreSQL. This file is considered the oldest that the system requires. The script also checks if any replication slots require specific WAL files, as subscribers lagging behind might need these files to catch up.
> 
> The script retrieves a list of all WAL files required by existing replication slots. It then applies the following logic: If a file is required by a replication slot, it is not deleted. Otherwise, if the file is not used by a replication slot and is older than the oldest WAL file currently in use by the PostgreSQL system, it is deleted using pg_cleanup archive command. For context, archiving is enabled, and all WAL files from pg_wal/* are copied to an archive/ directory for recovery and replication purposes.
> 
> PostgreSQL 16 also offers the recovery_command = 'command' option. This allows PostgreSQL to retrieve WAL file information from the archive/ directory if a WAL file is deleted from pg_wal before being consumed by a subscriber.
> 
> Please feel free to contact me if you have any questions about this concept or if you'd like assistance with your specific use case. Happy coding!

@spreeker @thangamani-arun
